File: 04.Simulation模拟/leetcode_2751_survivedRobotsHealths.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <stack>
using namespace std;

class Solution {
public:
    vector<int> survivedRobotsHealths(vector<int>& positions, vector<int>& healths, string directions) {
        int n = positions.size();
        vector<int> indexs;
        for (int i=0; i<n; i++){
            indexs.push_back(i);
        }
        sort(indexs.begin(), indexs.end(), [&](int& a, int& b){
            return positions[a]<positions[b];
        });

        stack<int> st;
        for (int& i : indexs){
            if (directions[i] == 'R'){
                st.push(i);
                continue;
            }
            while (!st.empty()){
                int j = st.top();
                if (healths[i] < healths[j]){
                    healths[i] = 0;
                    healths[j] -= 1;
                    break;
                }
                if (healths[i] == healths[j]){
                    healths[i] = 0;
                    healths[j] = 0;
                    st.pop();
                    break;
                }
                healths[i] -= 1;
                healths[j] = 0;
                st.pop();
            }
        }
        vector<int> res;
        for (int i : healths){
            if (i>0){
                res.push_back(i);
            }
        }
        return res;
    }
};

int main(){
    Solution s;
    vector<int> positions = {5,4,3,2,1}, healths = {2,17,9,15,10};
    string directions = "RRRRR";
    vector<int> res = s.survivedRobotsHealths(positions, healths, directions);
    for (int i : res){
        cout << i << " ";
    }
    cout << endl;
}

File: 04.Simulation模拟/leetcode_2751_survivedRobotsHealths.py
from typing import List

class Solution:
    def survivedRobotsHealths(self, positions: List[int], healths: List[int], directions: str) -> List[int]:
        n = len(positions)
        indexs = sorted(range(n), key = lambda i:positions[i])
        # 用索引数组可以在不改变原数组的情况下按照位置从小到大循环
        
        st = []
        for i in indexs:
            if directions[i] == 'R':
                st.append(i)
                continue

            while st:
                # 直接模拟三种情况
                j = st[-1]
                if healths[i] < healths[j]:
                    healths[i] = 0
                    healths[j] -= 1
                    break
                if healths[i] == healths[j]:
                    healths[i] = 0
                    healths[j] = 0
                    st.pop()
                    break
                healths[i] -= 1
                healths[j] = 0
                st.pop()
        
        res = []
        for i in healths:
            if i>0:
                res.append(i)
        return res

if __name__ == "__main__":
    s = Solution()
    positions = [5,4,3,2,1]
    healths = [2,17,9,15,10]
    directions = "RRRRR"
    print(s.survivedRobotsHealths(positions, healths, directions))
